src/operator/communication/gather.rs

```diff
@@ -0,0 +1,423 @@
+    }
+}
+
+struct GatherConsumer<T> {
+    queues: Box<[Consumer<T>]>,
+    ready_callback: NotifyCallback,
+    location: &'static Location<'static>,
+}
+
+impl<T> GatherConsumer<T> {
+    const fn new(
+        queues: Box<[Consumer<T>]>,
+        ready_callback: NotifyCallback,
+        location: &'static Location<'static>,
+    ) -> Self {
+        Self {
+            queues,
+            ready_callback,
+            location,
+        }
+    }
+
+    fn all_consumers_full(&self) -> bool {
+        !self.queues.iter().any(Consumer::is_empty)
+    }
+}
+
+impl<T: 'static> Operator for GatherConsumer<T> {
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("GatherConsumer")
+    }
+
+    fn location(&self) -> OperatorLocation {
+        Some(self.location)
+    }
+
+    fn is_async(&self) -> bool {
+        true
+    }
+
+    fn register_ready_callback<F>(&mut self, callback: F)
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        let _ = self.ready_callback.set(Box::new(callback));
+    }
+
+    fn ready(&self) -> bool {
+        self.all_consumers_full()
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<T> SourceOperator<Spine<T>> for GatherConsumer<T>
+where
+    T: Batch + 'static,
+    Spine<T>: Trace<Batch = T>,
+{
+    fn eval(&mut self) -> Spine<T> {
+        debug_assert!(self.all_consumers_full());
+
+        let mut spine = Spine::new(None);
+        for consumer in self.queues.iter_mut() {
+            let batch = *consumer
+                .pop()
+                .expect("GatherConsumer popped from empty channel");
+            spine.insert(batch);
+        }
+
+        spine
+    }
+}
+
+/// The consumer half of the gather operator that's given to all
+/// the workers who aren't the target of the gather, simply yields
+/// an empty trace on each clock cycle
+struct EmptyGatherConsumer<T> {
+    location: &'static Location<'static>,
+    __type: PhantomData<T>,
+}
+
+impl<T> EmptyGatherConsumer<T> {
+    const fn new(location: &'static Location<'static>) -> Self {
+        Self {
+            location,
+            __type: PhantomData,
+        }
+    }
+}
+
+impl<T: 'static> Operator for EmptyGatherConsumer<T> {
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("GatherConsumer")
+    }
+
+    fn location(&self) -> OperatorLocation {
+        Some(self.location)
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<T> SourceOperator<Spine<T>> for EmptyGatherConsumer<T>
+where
+    T: Batch + 'static,
+    Spine<T>: Trace<Batch = T>,
+{
+    fn eval(&mut self) -> Spine<T> {
+        Default::default()
+    }
+}
+
+/// A very specialized oneshot single producer, single consumer channel
+///
+/// # Safety
+///
+/// Requires external synchronization via circuit epochs, the [`Producer`]
+/// associated with a channel should only be pushed to once every clock cycle.
+///
+struct OneshotSpsc<T> {
```

Comment:
Can we replace mailboxes of type `Mutex<Option<T>>` in `exchange.rs` with this?

Reply:
Yah, I think so